Mathlib/Topology/DiscreteSubset.lean

```diff
@@ -104,4 +104,41 @@ def Filter.codiscrete (X : Type*) [TopologicalSpace X] : Filter X where
     simp_rw [← isClosed_compl_iff, isClosed_and_discrete_iff] at hU hV ⊢
     exact fun x ↦ compl_inter U V ▸ sup_principal ▸ disjoint_sup_right.mpr ⟨hU x, hV x⟩
 
+lemma mem_codiscrete {S : Set X} :
+    S ∈ codiscrete X ↔ IsOpen S ∧ DiscreteTopology ↑Sᶜ := Iff.rfl
+
+lemma mem_codiscrete' {S : Set X} :
+    S ∈ codiscrete X ↔ ∀ x, Disjoint (𝓝[≠] x) (𝓟 Sᶜ) := by
+  rw [mem_codiscrete, ← isClosed_compl_iff, isClosed_and_discrete_iff]
+
+lemma mem_codiscrete_subtype {S : Set X} {U : Set S} :
+    U ∈ codiscrete S ↔ ∀ x ∈ S, ¬AccPt x (𝓟 (Subtype.val '' Uᶜ)) := by
```

Comment:
Is this situation really coming up that frequently? We normally try to avoid situations like `{S : Set X} {U : Set S}`.

Reply:
I don't know how common this is in general, but when working with analytic and meromorphic functions on a set `S` I believe that `codiscrete S` is the best way to state "almost everywhere" (see #14820, for example).

Reply:
I guess I see a number of issues here:

1. I think the *definition* of `Filter.codiscrete` should be migrated to `mem_codiscrete'`, as it plays better with our API, and avoids a `CoeSort` in the definition.
2. It's weird that you phrase `mem_codiscrete'` with `Disjoint` but `mem_codiscrete_subtype` with `¬ AccPt`
3. This is the big one. I would prefer that we *almost never* talk about `U ∈ codiscrete S` where `{S : Set X} {U : Set S}`. Instead, I think we should create a *new* definition `codiscreteWithin`, much like `nhdsWithin`. I will describe this now.

The definition of `t ∈ codiscrete X` is (equivalent to) `Disjoint (⨆ x : X, 𝓝[≠] x) (𝓟 tᶜ)`. I propose that we have, for `s : Set X`, `codiscreteWithin s X` where `t ∈ codiscreteWithin s X` means `Disjoint (⨆ x ∈ s, 𝓝[≠] x) (𝓟 tᶜ)`. Note, I have not thought about whether there is a nice filter-level description of this (as with `𝓝[s] x := 𝓝 x ⊓ (𝓟 s)`).

Note that if `t : Set s`, then `t ∈ codiscrete s ↔ (Subtype.val '' tᶜ) ∈ codiscreteWithin s X`. Probably all of this should be discussed on Zulip. But the advantage of my approach is that it allows everything to stay on the top level.

Reply:
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/codiscrete.20within.20a.20set

Reply:
@Command-Master given the (albeit small) consensus from Zulip, can you please implement the following suggestions? (you may split this into multiple PRs to make review easier.)

1. (Optional, but would be my preference if it's not too much work, and it makes some of what follows slightly easier too) redefine `Filter.codiscrete X : Filter X := ⨆ x : X, 𝓝[≠] x`
2. Define for `s : Set X`, a new `Filter.codiscreteWithin s X : Filter X := ⨆ x ∈ s, 𝓝[≠] x`.
3. Develop the basic API for this new filter (e.g., `codiscreteWithin Set.univ X = codiscrete X`, or more generally, I think `(codiscreteWithin · X)` can be made into a `CompleteLatticeHom` pretty easily (if I'm not mistaken).

In the above you are welcome to include the equivalent conditions in terms of subtypes, but I think we should try to avoid them when / where possible.

Reply:
I'm not sure if `⨆ x ∈ s, 𝓝[≠] x` is a correct definition for `codiscreteWithin` - using it, `S ∈ codiscreteWithin T` means that `Sᶜ` has no accumulation points within `T`, but I think the correct definition should mean that `T \ S` has no accumulation points within `T` (this also agrees with `codiscrete` on the subtype), that is `⨆ x ∈ S, 𝓝[S \ {x}] x`.
Also, shouldn't `X` be implicit in `codiscreteWithin`? Finally, what do you think about having `codiscrete X = codiscreteWithin Set.univ X` directly as the definition?

Reply:
Sure, if that's the version equivalent to the subtype for `codiscrete`, then let's go with that. Sorry I got it wrong before.

Yes `X` should be implicit.

I'm tempted to keep them separate, since the definition is cleaner, but perhaps an `abbrev` with the right API lemmas is okay. That's a question for another reviewer / Zulip (but you don't have to wait forever on Zulip, sorry it took so long before).